pkg/ebpf/benchmark_test.go

```diff
@@ -0,0 +1,156 @@
+}
+
+func setBPFStatsCollection(b *testing.B, enable bool) error {
+	if currKernelVersion < stringToKernelCode("5.1.0") {
+		b.Skip("kernel version must be at least 5.1.0 for kernel-level eBPF stats collection")
+		return nil
+	}
+
+	flag := 0
+	if enable {
+		flag = 1
+	}
+	cmd := exec.Command("sysctl", "-w", fmt.Sprintf("kernel.bpf_stats_enabled=%d", flag))
+	if out, err := cmd.CombinedOutput(); err != nil {
+		return fmt.Errorf("error setting sysctl kernel.bpf_stats_enabled: %s %w", out, err)
+	}
+	return nil
+}
+
+func collectBPFToolStats(progID uint32) (*bpftoolStats, error) {
+	cmd := exec.Command(bpfToolPath, "--json", "prog", "show", "id", fmt.Sprintf("%d", progID))
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		return nil, fmt.Errorf("error collecting bpf stats: %s %w", out, err)
+	}
+	stats := bpftoolStats{}
+	err = json.Unmarshal(out, &stats)
+	if err != nil {
+		return nil, fmt.Errorf("error unmarshaling json from bpftool: %s", err)
+	}
+	return &stats, nil
+}
+
+func getAllProbeStats(t *Tracer) (map[string]*bpftoolStats, error) {
+	ids, err := t.getProbeProgramIDs()
+	if err != nil {
+		return nil, err
+	}
+
+	allStats := make(map[string]*bpftoolStats, len(ids))
+	for name, id := range ids {
+		stats, err := collectBPFToolStats(id)
+		if err != nil {
+			return nil, err
+		}
+		allStats[name] = stats
+	}
+	return allStats, nil
+}
+
+func RunEBPFBenchmark(b *testing.B, fn func(*testing.B)) {
+	readVars.Do(readGlobalVars)
+
+	err := setBPFStatsCollection(b, true)
+	if err != nil {
+		b.Fatal(err)
+	}
+	defer setBPFStatsCollection(b, false)
+
+	// Enable BPF-based system probe
+	t, err := NewTracer(NewDefaultConfig())
+	if err != nil {
+		b.Fatal(err)
+	}
+	defer t.Stop()
+
+	var prs map[string]*testing.BenchmarkResult
+	b.Run("probes", func(b *testing.B) {
+		baseline, err := getAllProbeStats(t)
+		require.NoError(b, err)
+
+		b.ResetTimer()
+		fn(b)
+		b.StopTimer()
+
+		post, err := getAllProbeStats(t)
+		require.NoError(b, err)
+
+		// override outer variable here so we only report on the last run of results
+		prs = make(map[string]*testing.BenchmarkResult, len(baseline))
+		for probeName, base := range baseline {
+			p := post[probeName]
+			runTime := p.RunTimeNS - base.RunTimeNS
+			runCount := p.RunCnt - base.RunCnt
+			prs[probeName] = &testing.BenchmarkResult{
+				N: int(runCount),
+				T: time.Duration(runTime) * time.Nanosecond,
+			}
+		}
+	})
+	fmt.Print(prettyPrintEBPFResults(b.Name(), prs))
+}
+
+func prettyPrintEBPFResults(name string, probeResults map[string]*testing.BenchmarkResult) string {
+	maxLen := 0
+	var probeNames []string
+	for pName := range probeResults {
+		if len(pName) > maxLen {
+			maxLen = len(pName)
+		}
+		probeNames = append(probeNames, pName)
+	}
+	sort.Strings(probeNames)
+	buf := new(strings.Builder)
+	for _, probeName := range probeNames {
+		pr := probeResults[probeName]
+		if pr.N == 0 {
+			continue
+		}
+		fmt.Fprintf(buf, "%s/%-*s\t%s\n", name, maxLen, probeName, pr.String())
```

Comment:
is this format required by benchstat ?

Reply:
The `%-*s` allows us to specify a variable-based width specifier, which makes it so the columns line up nicely (as is done in stdlib). Calling `BenchmarkResult.String()` gets us the results (without the name) in the go benchmark format.